### ISSUE TITLE
As a Game Designer I need to specify turn interval in months/years

### ISSUE THREAD

**Issue by IanMayo** [serge-web/serge#991]
Serge currently allows game turn lengths to be specified using days, hours, minutes, seconds:
![image](https://user-images.githubusercontent.com/1108513/129016115-c59b0f9b-feb3-4c25-abb3-630395456901.png)

For larger scale wargames, the interval could be in months or years, such as each turn spanning a two-month period (Jan21-Feb21, then Mar21-Apr21, then May21-Jun-12, etc). Given the changing length of both of these items (in terms of days), we need an alternate way of specifying game turn length.

So, I _think_ we need a range of turn length strategies:
* days/hours/minutes
* months
* years

Months and years would just take an integer value, using as appropriate in the turn time incrementing.

I _think_ we need a two-step UI.  First select the turn mode, then enter the value.

Here is what the data structure will look like, I guess:
```
export interface MillisecondTurns {
  millis: number
}

export interface MonthTurns {
  months: bigInt
}

export interface YearTurns {
 years: bigInt
}

export type GameTurnLength = MillisecondTurns | MonthTurns | YearTurns

```

Well, to be backward compatible with our current data, it would be:
```
export type GameTurnLength = number | MillisecondTurns | MonthTurns | YearTurns
```

**Comment by IanMayo** [comment]
Hello @ShendMikullovci 
this is a feature that overlaps with your other design task (#954)

So far, the turn-period has been expressed in days, hours, or minutes.  All of these can be represented by a single number (milliseconds) in the software, or via a `DD HH:MM:SS` masked text control.

But, in this game they want the turn to move forward 2 months. So, they expect it to step from 1st January to 1st March - **not by 60 days**.  The time period represented by two months changes - so we need a different way of representing that.  While we add `months`, we should probably also support `years`

So that's 3 different ways of expressing the turn size:
* Days/Hours/Minutes
* Months
* Years

I _guess_ we could support this through a drop-down to select the type, then a suitable text field to provide the detail.  Or, a control with three tabs - one for each style. Each tab would contain a suitable edit control - and we'd store the value of the selected tab.

Aah, once the design for this is complete, we can feed it back into #954 

(I've separated this out because this feature would be useful for the current wargame)

**Comment by ShendMikullovci** [comment]
After our talk, and our joint affinity over the Tabs idea, I decided to give the design a try. This is how the tabs look: 
![Format Tab](https://user-images.githubusercontent.com/85491794/133891793-91615945-95d1-46c1-8afd-d511a3a819c5.png)

DHM Stands for Days/Hours/Minutes. 

Here DHM is the selected mode as you can see by the highlighted DHM rectangle. “Turn Duration” represents the length of the turn in the primary time format, in this example, it would be 1 Day. The format selected to display the desired view of the mode is DAYSDAYS:HOURSHOURS:MINUTESMINUTES. 

Once we collapse the “Format” drop-down, we see the option to display different formats. “YY/MM/DD HH:MM” is dimmed because it logically seems too long of a time length to display time in years, if we have selected the DHM mode. If we need to account for Months and/or Years, we click on “Months” or “Years” mode to do so.

Refer to Issue #954 to view it in the “Game Overview” window.

Let me know what you think!

**Comment by IanMayo** [comment]
Thanks @ShendMikullovci - the single integer value is ok for `months` or `years`.  But, for DHM we need a text field with `DD HH MM SS`.
 So, a 2 day step would be:
`02 00 00 00`
Or a 30 minute step would be:
`00 00 30 00`
And so on.

**Comment by ShendMikullovci** [comment]
Noted! Here's how it looks now:
![Format Tab](https://user-images.githubusercontent.com/85491794/133906490-1cf237ef-64d2-415a-8679-8ad711b34351.png)

![Game Overview Optimization 3 Tab Addition](https://user-images.githubusercontent.com/85491794/133906494-7305618a-a64a-43b2-9fbf-0eb23bedf1fd.png)

**Comment by IanMayo** [comment]
Yes, great. The labels (like "Turn Duration") should be _almost_ alongside the control they refers to.

I _don't think_ we need "dark mode" for the new panel.  It's not all that different to the other content.

**Comment by ShendMikullovci** [comment]
Noted. Here’s how it looks:

![Game Overview Optimization 3 Tab Addition](https://user-images.githubusercontent.com/85491794/133982627-0ac8d96e-a865-4f5b-80b8-26c9ea65b353.png)
![Format Tab](https://user-images.githubusercontent.com/85491794/133982633-4fda05a6-b3bc-4d5a-acc9-1a3e7900ec5c.png)

Kept the darkened tab on the DHM option to emphasize the selected mode. Would you like for me to do the designs on the “Months” and “Years” tabs as well? Will this suffice for the developing team?

**Comment by IanMayo** [comment]
Yes- it's right to highlight the selected tab.  In the libraries we use the defaults are usually quite sensible - so we normally tend to "go with the flow" and accept the default colors.

Aah - yes, it would be useful to see `Months`, since it's a simpler control (just single number).

**Comment by ShendMikullovci** [comment]
Great. There we are: 

![Format Tab Month Mode](https://user-images.githubusercontent.com/85491794/133985619-6735d36d-9af0-4b37-acb6-0ab607aa76b1.png)

![Game Overview Optimization 3 Tab Addition Month Mode](https://user-images.githubusercontent.com/85491794/133985621-8d1c78d0-72e2-4621-af4f-a540954ea400.png)

**Comment by IanMayo** [comment]
@ShendMikullovci - we're slowing getting on top of the new feature workload, and can start working on some of the lower priority bits.

I'd appreciate you giving an update to this UI, please.  I have these observations:
1. for the `In-Game time`  we won't need the "Years" tab, since a year is just a multiple of "Months".
2. Could you use styling to show which are editable text boxes
3. I don't believe we need any of the underlining

**Comment by ShendMikullovci** [comment]
Gave it a fresher look that matches what we have in the admin pages so far: 

![Overview 1](https://user-images.githubusercontent.com/85491794/179371508-912c5c41-d2bb-4f73-b7a1-c3e5a418d056.png)

![Overview 2](https://user-images.githubusercontent.com/85491794/179371512-4065d3ce-36a8-43ad-a0b6-86a06de4dbd4.png)

While this layout occupies more space, we'd conserve uniformity here by having features displayed vertically as we do in other admin page tabs. Notice I also increased the tabs' sizes since they are the key navigational buttons to click on the admin page. Notice that I also gave the Save button a lighter blue for visual purposes.

Editing the features would lead the user to select options through drop-downs or to manually insert them, depending on the item to be edited: 

![Editing Overview](https://user-images.githubusercontent.com/85491794/179371668-c8aeec62-29d1-4697-9288-db9d531fd0b7.png)

![Editing Overview 2](https://user-images.githubusercontent.com/85491794/179371672-ab0ec92a-f248-45b8-b92a-df25681d9430.png)

![Editing Overview 3](https://user-images.githubusercontent.com/85491794/179371676-beabceaa-1864-4225-a5de-99c3d6e33f41.png)

There's the possibility to have radio buttons instead of the drop-downs to reduce click steps, but I decided to stick to the drop-downs to maintain the same logic we use on other tabs. Let me know if you'd like to see radio buttons instead.

**Comment by IanMayo** [comment]
I see the consistency.  But, as the window gets wider, the label gets further from the control it relates to.  

I don't think the "trash" icon has a meaning in this context, since the fields are always present.  Hmm, if it means "reset to default" then the `revert` arrow may be more valid.

While I recognise the meaning of "Start time" and "Current time" - I can't recall what "Game Time" represents. Would you mind checking back, to investigate please? (possibly back in #954 ).

**Comment by ShendMikullovci** [comment]
> I see the consistency. But, as the window gets wider, the label gets further from the control it relates to.

Given the page is not overloaded with options, the distance shouldn't be an issue. I could of course shrink things and eventually split them into two different windows, but it would break from our current design logic and wouldn't really be too intuitive in my opinion. Let me know if you want me to try.

> I don't think the "trash" icon has a meaning in this context, since the fields are always present. Hmm, if it means "reset to default" then the `revert` arrow may be more valid.

Yup, good thinking. Replaced it: 

![Overview new](https://user-images.githubusercontent.com/85491794/179616764-1affd6d8-b858-4e7c-be93-d052218d10cd.png)


> While I recognise the meaning of "Start time" and "Current time" - I can't recall what "Game Time" represents. Would you mind checking back, to investigate please? (possibly back in #954 ).

It's a sub-header that we used to replace Display Time, containing 4 options ("Time of Start of Turn", "Time of End of Turn", "Turn Start and End Times", "Do Not Display Times"). If it's irrelevant, we could replace it with your idea in #1400 : 

> One observation - we need the "time sent" in the message. Our current implementation may not be enough. Some wargames span multiple days - so we'd need to display date and time for each message. Hmm, this should probably be a setting in the "Overview" panel - since the Game Designer will know if the game is to span whole days or not.

**Comment by IanMayo** [comment]
>  I could of course shrink things and eventually split them into two different windows

Having two/more columns of properties means more can be reviewed at once, without scrolling.  So, I prefer your general strategy of Sept 2021, but with updated styling/presentation.

> It's a sub-header that we used to replace Display Time, containing 4 options ("Time of Start of Turn", "Time of End of Turn", "Turn Start and End Times", "Do Not Display Times"). If it's irrelevant, we could replace it with your idea in https://github.com/serge-web/serge/issues/1400 :

Brilliant! Over the weekend I was thinking we needed something like this.  Let's rename "Game Time" to "Display format".

**Comment by ShendMikullovci** [comment]
Here's one potential look: 

![Overview new](https://user-images.githubusercontent.com/85491794/179863172-e80782c1-7b92-40b5-9060-b32db8110440.png)

Renamed "Game Time" to "Display Format". Got me thinking though, isn't the DHM/Months feature just an extension of the format for time we display? If so, shouldn't "Display Format" be down there by DHM/Months, or included in, or simply conjoined to them somehow?

**Comment by IanMayo** [comment]
Layout works for me :-)

Aah, we may be able to re-arrange a little more logically.  The "DHM/Months" box is how far forward we move at each Turn.  So, we could move "Real World Time" beneath "In-game time", and move "DHM"  under "Turns".

**Comment by ShendMikullovci** [comment]
Good idea! 

![Overview new](https://user-images.githubusercontent.com/85491794/180096809-285c93f1-56f5-4834-92b3-00eec3eadd35.png)

**Comment by IanMayo** [comment]
Not quite - the DHM relates to the turns, so I think we should give DHM a `Turn length` title, and lay out like this:
```
In-game time         Turn Style
Real World Time      Turn Length
```

**Comment by ShendMikullovci** [comment]
Something like this?:

![Overview](https://user-images.githubusercontent.com/85491794/180753956-5a4d65bf-357e-4524-bbb7-df9be03e4a37.png)

Added the "access codes" and "hide forces" to the top in smaller fonts, but we can relocate them. Putting them on the bottom right or left felt a little imbalanced.

I'm starting to like the toggle option for DHM/Months less and less and considering turning into a tab set up. What do you think?

**Comment by IanMayo** [comment]
> I'm starting to like the toggle option for DHM/Months less and less and considering turning into a tab set up. What do you think?

Yes, I think that putting the two turn length options into tabs with the relevant controls will "connect" them together.

> Added the "access codes" and "hide forces" to the top in smaller fonts, but we can relocate them. Putting them on the bottom right or left felt a little imbalanced.

How about adding them to the "Game Description and Objectives" box?  They seem a bit "homeless" currently - floating around without a containing box.

**Comment by ShendMikullovci** [comment]
https://user-images.githubusercontent.com/85491794/180761158-e4e4b815-a382-4b23-8685-614db3b44580.mp4

>How about adding them to the "Game Description and Objectives" box? They seem a bit "homeless" currently - floating around without a containing box.

Added them below the Objectives text box.

**Comment by IanMayo** [comment]
Yes, tabs work well for me.
